Ical.Net/Evaluation/EventEvaluator.cs

```diff
@@ -52,52 +52,38 @@ public override IEnumerable<Period> Evaluate(CalDateTime referenceDate, CalDateT
     /// </summary>
     /// <param name="period">The period where <see cref="Period.EndTime"/> will be set.</param>
     /// <returns>Returns the <paramref name="period"/> with <see cref="Period.EndTime"/> and exact <see cref="Period.Duration"/> set.</returns>
-    private Period WithEndTime(Period period)
+    private Period WithFinalDuration(Period period)
     {
         try
         {
             /*
-               We use a time span to calculate the end time of the event.
-               It evaluates the event's definition of DtStart and either DtEnd or Duration.
+               The period's Duration evaluates the event's definition of DtStart
+               and Duration, or the timespan from DtStart to DtEnd.
 
                The time span is used, because the period end time gets the same timezone as the event end time.
                This ensures that the end time is correct, even for DST transitions.
 
                The exact duration is calculated from the zoned end time and the zoned start time,
                and it may differ from the time span added to the period start time.
              */
-            var tsToAdd = CalendarEvent.EffectiveDuration;
 
-            CalDateTime endTime;
-            if (tsToAdd.IsZero)
+            // The preliminary duration was set in a previous evaluation step
+            var duration = period.Duration;
+
+            if (duration == null)
             {
-                // For a zero-duration event, the end time is the same as the start time.
-                endTime = period.StartTime;
+                duration = CalendarEvent.EffectiveDuration;
             }
-            else
-            {
-                // Calculate the end time of the event as a DateTime
-                var endDt = period.StartTime.Add(tsToAdd);
-                if ((CalendarEvent.End is { } end) && (end.TzId != period.StartTime.TzId) && (end.TzId is { } tzid))
-                {
-                    // Ensure the end time has the same timezone as the event end time.
-                    endDt = endDt.ToTimeZone(tzid);
-                }
 
-                endTime = endDt;
-            }
+            var newPeriod = new Period(
+                start: period.StartTime,
+                duration: duration.Value);
 
-            // Return the Period object with the calculated end time.
-            // Only EndTime is relevant for further processing,
-            // so we have to set it.
-            // If the period duration is not null here, it is an RDATE period
-            // and has priority over the calculated end time.
+            // Ensure that adding the duration to the start time
+            // results in a representable date/time
+            _ = newPeriod.EffectiveEndTime;
```

Comment:
In which scenario would we get a non-representable date/time. And if we did, why would we throw here? Wouldn't it be one of the strengths of this approach that we can let the user choose, how to deal with this case? Maybe they don't need the EndTime at all, so why throw then?

Reply:
Can be caused by unlimited GetOccurrences, see #785

Reply:
Agree